File: 05-Algorithms/code/typescript/src/maxFlow.ts
import { AlgoError } from "./errors.js";
import { checkVertexInRange, type WeightedEdge } from "./graphCommon.js";

/**
 * Maximum flow from `s` to `t` via Edmonds-Karp: repeatedly find an
 * augmenting path with BFS over the residual graph (always the
 * fewest-hops path among residual-capacity-positive edges, exploring
 * neighbors in ascending vertex-id order for determinism), and push the
 * bottleneck capacity along it. O(V * E^2). Edge list is directed
 * (u -> v with the given capacity); parallel edges between the same pair
 * are summed into one residual-capacity entry. Throws AlgoError("invalid")
 * for negative capacities.
 */
export function maxFlow(n: number, edges: WeightedEdge[], s: number, t: number): number {
  checkVertexInRange(n, s);
  checkVertexInRange(n, t);
  const capacity: number[][] = Array.from({ length: n }, () => new Array(n).fill(0));
  for (const [u, v, c] of edges) {
    if (c < 0) throw new AlgoError("invalid", "maxFlow requires non-negative capacities");
    capacity[u][v] += c;
  }

  let totalFlow = 0;

  while (true) {
    const parent = new Array(n).fill(-1);
    parent[s] = s;
    const queue = [s];
    let head = 0;
    while (head < queue.length && parent[t] === -1) {
      const u = queue[head++];
      for (let v = 0; v < n; v++) {
        if (parent[v] === -1 && capacity[u][v] > 0) {
          parent[v] = u;
          queue.push(v);
        }
      }
    }
    if (parent[t] === -1) break;

    let bottleneck = Infinity;
    for (let v = t; v !== s; v = parent[v]) {
      bottleneck = Math.min(bottleneck, capacity[parent[v]][v]);
    }
    for (let v = t; v !== s; v = parent[v]) {
      capacity[parent[v]][v] -= bottleneck;
      capacity[v][parent[v]] += bottleneck;
    }
    totalFlow += bottleneck;
  }

  return totalFlow;
}

File: 05-Algorithms/code/python/seb_alg/max_flow.py
"""Mirrors `typescript/src/maxFlow.ts` exactly."""

from __future__ import annotations

from typing import List, Sequence

from .errors import AlgoError
from .graph_common import check_vertex_in_range


def max_flow(n: int, edges: Sequence[Sequence[int]], s: int, t: int) -> int:
    """Maximum flow from s to t via Edmonds-Karp: BFS augmenting paths over
    the residual graph, exploring neighbors in ascending id order. O(V*E^2).
    Parallel edges are summed. Raises AlgoError("invalid") for negative
    capacities.
    """
    check_vertex_in_range(n, s)
    check_vertex_in_range(n, t)
    capacity: List[List[int]] = [[0] * n for _ in range(n)]
    for u, v, c in edges:
        if c < 0:
            raise AlgoError("invalid", "maxFlow requires non-negative capacities")
        capacity[u][v] += c

    total_flow = 0

    while True:
        parent = [-1] * n
        parent[s] = s
        queue = [s]
        head = 0
        while head < len(queue) and parent[t] == -1:
            u = queue[head]
            head += 1
            for v in range(n):
                if parent[v] == -1 and capacity[u][v] > 0:
                    parent[v] = u
                    queue.append(v)
        if parent[t] == -1:
            break

        bottleneck = float("inf")
        v = t
        while v != s:
            bottleneck = min(bottleneck, capacity[parent[v]][v])
            v = parent[v]
        v = t
        while v != s:
            capacity[parent[v]][v] -= bottleneck
            capacity[v][parent[v]] += bottleneck
            v = parent[v]
        total_flow += bottleneck

    return int(total_flow)
